File: zZ/java/multiply-strings.java
/**
 * Multiply Strings
 *  
 * Given two numbers represented as strings, return multiplication of the numbers as a string.
 * 
 * Note: The numbers can be arbitrarily large and are non-negative.
 */

public class Solution {
    public String multiply(String num1, String num2) {
        if (num1 == null || num2 == null) {
            throw new NullPointerException();
        }

        // multiply
        int num1Length = num1.length();
        int num2Length = num2.length();
        int[] multiply = new int[num1Length + num2Length];
        for (int i = 0; i < num1Length; ++i) {
            for (int j = 0; j < num2Length; ++j) {
                multiply[num1Length + num2Length - 2 - i - j] += (num1.charAt(i) - '0') * (num2.charAt(j) - '0');
            }
        }

        // carry
        int carry = 0;
        for (int i = 0; i < num1Length + num2Length; ++i) {
            multiply[i] += carry;
            carry = multiply[i] / 10;
            multiply[i] %= 10;
        }

        // number is zero
        int i = num1Length + num2Length - 1;
        while (multiply[i--] == 0);
        if (i < 0) {
            return "0";
        }

        // number to string
        StringBuilder stringBuilder = new StringBuilder();
        while (i >= 0) {
            stringBuilder.append(multiply[i--] + '0');
        }
        return stringBuilder.toString();
    }
}

File: data-structure-and-algorithm/typescript/leetcode/src/multiply-strings.js
/*
Given two non-negative integers num1 and num2 represented as strings, return the product of num1 and num2, also represented as a string.

Note: You must not use any built-in BigInteger library or convert the inputs to integer directly.

 

Example 1:

Input: num1 = "2", num2 = "3"
Output: "6"
Example 2:

Input: num1 = "123", num2 = "456"
Output: "56088"
 

Constraints:

1 <= num1.length, num2.length <= 200
num1 and num2 consist of digits only.
Both num1 and num2 do not contain any leading zero, except the number 0 itself.
*/

function multiply(num1, num2) {
    const result = new Array(num1.length + num2.length).fill(0);

    for (let i = num1.length - 1; i >= 0; --i) {
        for (let j = num2.length - 1; j >= 0; --j) {
            const multi = num1[i] * num2[j];
            const di = i + j + 1, ci = i + j;
            const sum = multi + result[di];
            result[ci] += Math.trunc(sum / 10);
            result[di] = sum % 10;
        }
    }

    return BigInt(result.join("")).toString();
}